File: 프로그래머스/Python/0단계/다항식 더하기.py
# polynomial이 매개변수로 주어질 때, 동류항끼리 더한 결괏값을 문자열로 return 하도록 solution 함수를 완성해보세요.
# 같은 식이라면 가장 짧은 수식을 return 합니다.


# 내 풀이
def solution(polynomial):
    polynomial = polynomial.replace(' ','')
    arr1 = polynomial.split('+')
    a = []
    b = []
    for i in arr1:
        if 'x' in i:
            if i == 'x':
                a.append('1')
            else:
                a.append(i[0:len(i)-1])
        else: 
            b.append(i)
    a = sum(list(map(int,a)))
    b = sum(list(map(int,b)))
    if a == 0:
        return str(b)
    elif a == 1 and b == 0:
        return 'x'
    elif a == 1:
        return 'x'+' + '+str(b)
    elif b == 0:
        return str(a)+'x'
    else:
        return str(a)+'x'+' + '+str(b)


# 다른사람 풀이
def solution(polynomial):
    xnum = 0   # x계수
    const = 0  # 정수
    for c in polynomial.split(' + '): # +와 앞뒤 공백을 기준으로 문자열 분리
        if c.isdigit():  # 문자열도 isdigit()을 쓸 수 있다. 
            const+=int(c)  # c가 숫자면, const에 c더함
        else:
            xnum = xnum+1 if c=='x' else xnum+int(c[:-1])  # c가 x면 1을 더하고, 아니면 x를 제외한 부분까지 더함
    if xnum == 0:  # x의 계수가 0이면
        return str(const)
    elif xnum==1:  # x의 계수가 1이면
        return 'x + '+str(const) if const!=0 else 'x'  # 숫자부분이 0이 아니면 앞처럼 출력, 아니면 뒤처럼 출력
    else:
        return f'{xnum}x + {const}' if const!=0 else f'{xnum}x'

File: 프로그래머스/JavaScript/0단계/다항식 더하기.js
// 내 풀이
function solution(polynomial) {
    const arr1 = [];
    const arr2 = [];

    for(let a of polynomial.split(' + ')){
        if (a.includes('x')) arr1.push((a==='x')?1:Number(a.slice(0,a.length-1)))
        else arr2.push(Number(a))
    }
    let sum_arr1 = arr1.reduce((sum,curr)=>sum+curr,0);
    let sum_arr2 = arr2.reduce((sum,curr)=>sum+curr,0);

    if (sum_arr1 === 0) return String(sum_arr2);
    else if (sum_arr1 === 1 && sum_arr2 === 0) return 'x'
    else if (sum_arr1 === 1) return "x"+" + "+String(sum_arr2);
    else if (sum_arr2 === 0) return String(sum_arr1)+'x';
    else return String(sum_arr1)+'x'+' + '+String(sum_arr2);
}



// 다른사람 풀이
function solution(polynomial) {
    const arr = polynomial.split(" + ");
    const xNum = arr
                .filter(n => n.includes("x"))
                .map(n => n.replace('x', '') || '1')
                .reduce((acc, cur) => acc + parseInt(cur, 10), 0);
    const num = arr
                .filter(n => !isNaN(n))
                .reduce((acc, cur) => acc + parseInt(cur, 10), 0);

    let answer = [];
    if(xNum) answer.push(`${xNum === 1 ? "" : xNum}x`);
    if(num) answer.push(num);

    return answer.join(" + ");
}
